src/python/pants/rules/core/lint.py

```diff
@@ -33,17 +27,14 @@ class Lint(Goal):
 
 
 @console_rule
-def lint(console: Console, targets: HydratedTargets) -> Lint:
+def lint(console: Console, targets: HydratedTargets, union_membership: UnionMembership) -> Lint:
   results = yield [
           Get(LintResult, FmtTarget, target.adaptor)
           for target in targets
-          # @union assumes that all targets passed implement the union, so we manually
-          # filter the targets we know do; this should probably no-op or log or something
-          # configurable for non-matching targets.
-          # We also would want to remove the workaround that filters adaptors which have a
-          # `sources` attribute.
-          # See https://github.com/pantsbuild/pants/issues/4535
-          if isinstance(target.adaptor, (PythonAppAdaptor, PythonTargetAdaptor, PythonTestsAdaptor, PythonBinaryAdaptor)) and hasattr(target.adaptor, "sources")
+           # We will want to remove the workaround that filters adaptors which have a `sources`
+           # attribute.
+           # See https://github.com/pantsbuild/pants/issues/4535
+          if union_membership.is_member(FmtTarget, target.adaptor) and hasattr(target.adaptor, "sources")
```

Comment:
Why do we in fact need the "sources" attr check? It seems unusual since we control which types pass the first check. That link is not particularly helpful as it doesn't directly address this question. Better to explain explicitly here.

Reply:
I agree that we probably don't need the sources attr check any more

Reply:
We do: I tried to remove it and it failed

Reply:
I will add a bit more context to the comment, but this is a bit mysterious to me too: I'm not sure why some of the targets in this union don't have sources; I just observe it is so.

Reply:
btw, the exception I'm getting when running, for instance:
```
./pants lint-v2 examples::
```
is this:
```
Exception message: 1 Exception encountered:
Computing Select((<pants.engine.console.Console object at 0x112215400>+OptionsBootstrapper(env_tuples=(('PANTS_DEV', '1'), ('PANTS_HOME', '/Users/pchevalier/.cache/pants')), bootstrap_args=(), args=('/Users/pchevalier/Documents/code/pants
/src/python/pants/bin/pants_loader.py', 'lint-v2', 'examples::'), config=_ChainedConfig(chained_configs=(<pants.option.config._SingleFileConfig object at 0x10832bcf8>,)))+Specs(dependencies=(DescendantAddresses(directory='examples'),), ma
tcher=SpecsMatcher(tags=(), exclude_patterns=()))), Lint)
  Computing Task(lint(), (<pants.engine.console.Console object at 0x112215400>+OptionsBootstrapper(env_tuples=(('PANTS_DEV', '1'), ('PANTS_HOME', '/Users/pchevalier/.cache/pants')), bootstrap_args=(), args=('/Users/pchevalier/Documents/co
de/pants/src/python/pants/bin/pants_loader.py', 'lint-v2', 'examples::'), config=_ChainedConfig(chained_configs=(<pants.option.config._SingleFileConfig object at 0x10832bcf8>,)))+Specs(dependencies=(DescendantAddresses(directory='examples
'),), matcher=SpecsMatcher(tags=(), exclude_patterns=()))), Lint, false)
    Computing Task(lint_with_black(), (OptionsBootstrapper(env_tuples=(('PANTS_DEV', '1'), ('PANTS_HOME', '/Users/pchevalier/.cache/pants')), bootstrap_args=(), args=('/Users/pchevalier/Documents/code/pants/src/python/pants/bin/pants_load
er.py', 'lint-v2', 'examples::'), config=_ChainedConfig(chained_configs=(<pants.option.config._SingleFileConfig object at 0x10832bcf8>,)))+PythonAppAdaptor(address=examples/src/python/example/hello/main:hello-app)), LintResult, true)
      Computing Task(get_black_input(), (OptionsBootstrapper(env_tuples=(('PANTS_DEV', '1'), ('PANTS_HOME', '/Users/pchevalier/.cache/pants')), bootstrap_args=(), args=('/Users/pchevalier/Documents/code/pants/src/python/pants/bin/pants_lo
ader.py', 'lint-v2', 'examples::'), config=_ChainedConfig(chained_configs=(<pants.option.config._SingleFileConfig object at 0x10832bcf8>,)))+PythonAppAdaptor(address=examples/src/python/example/hello/main:hello-app)), BlackInput, true)
        Throw('PythonAppAdaptor' object has no attribute 'sources')
          Traceback (most recent call last):
            File "/Users/pchevalier/Documents/code/pants/src/python/pants/engine/native.py", line 463, in extern_generator_send
              res = c.from_value(func[0]).send(c.from_value(arg[0]))
            File "/Users/pchevalier/Documents/code/pants/src/python/pants/backend/python/rules/python_fmt.py", line 64, in get_black_input
              sources_digest = target.sources.snapshot.directory_digest
            File "/Users/pchevalier/Documents/code/pants/src/python/pants/engine/struct.py", line 253, in __getattr__
              return object.__getattribute__(self, item)
          AttributeError: 'PythonAppAdaptor' object has no attribute 'sources'
```

Do you think `PythonAppAdaptor` may be excluded from the union? (I lack background knowledge to make that call)

Reply:
A `PythonApp` indeed does not support sources: can find in page for `python_app` on https://www.pantsbuild.org/build_dictionary.html